pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -404,9 +407,9 @@ public boolean hasNext() {
               values[i] = value;
             }
             PrimaryKey primaryKey = new PrimaryKey(values);
-            Object timeValue = columnToReaderMap.get(_timeColumnName).getValue(_docId);
+            Object timeValue = columnToReaderMap.get(_upsertComparisonColumn).getValue(_docId);
             Preconditions.checkArgument(timeValue instanceof Comparable, "time column shall be comparable");
-            long timestamp = IngestionUtils.extractTimeValue((Comparable) timeValue);
+            Comparable timestamp = IngestionUtils.extractTimeValueIfPossible((Comparable) timeValue);
```

Comment:
Directly put the value instead of calling `extractTimeValue()`. We can handle general comparable values